File: SphereDist.cpp
// SphereDist (equal distance distribution of vertexes on a sphere or a hemisphere).
// Thanks to Atsushi Yamaji (Kyoto Univ.).
// Copyright (C) 2020 Katayama Hirofumi MZ <katayama.hirofumi.mz@gmail.com>.
// License: MIT

#include "SphereDist.h"
#include <cstdlib>
#include <cstdio>

#ifndef M_PI
    #define M_PI 3.141592653589793
#endif

inline void vector_add(Tvector3& a, const Tvector3& b)
{
    a.x += b.x;
    a.y += b.y;
    a.z += b.z;
}

static int Number_of_points;                    // N
static int NP_on_hemisphere;                    // N/2
static std::vector<TSpherical_Coordinates> GSS; // GSS

static bool GenerateGSS(int N)
{
    int k;
    double hk;
    const int Min_N = 6;
    if (N < Min_N)
    {
        std::fprintf(stderr, "SphereDist: N must be >= %d.\n", Min_N);
        return false;
    }

    Number_of_points = N;
    GSS.resize(N + 1);

    GSS[1].lon = 0;
    GSS[1].col = M_PI;

    for (k = 2; k <= N - 1; ++k)
    {
        hk = -1.0 + 2.0 * (k - 1) / (N - 1);
        GSS[k].col = acos(hk);
        GSS[k].lon = GSS[k - 1].lon + 3.6 / sqrt(N) / sqrt(1.0 - hk * hk);
    }

    GSS[N].lon = 0;
    GSS[N].col = 0;

    return true;
}

static void RelocatePoints(void)
{
    const std::vector<int> five_points = { 0xFFFF, 2, 3, 5, 6, 7 };
    Tvector3 mean;
    int i;

    // relocate 1st point
    mean.x = mean.y = mean.z = 0.0;
    for (i = 1; i <= 5; ++i)
    {
        vector_add(mean, Spherical2Cartesian(GSS[five_points[i]]));
    }
    GSS[1] = Cartesian2Spherical(mean);

    // relocate Nth point
    mean.x = mean.y = mean.z = 0.0;
    for (i = 1; i <= 5; ++i)
    {
        vector_add(mean, Spherical2Cartesian(GSS[Number_of_points - five_points[i] + 1]));
    }
    GSS[Number_of_points] = Cartesian2Spherical(mean);
}

static int GetNP_on_hemisphere(void)
{
    int ith_point;

    NP_on_hemisphere = 0;
    for (ith_point = 1; ith_point <= Number_of_points; ++ith_point)
    {
        if (GSS[ith_point].col <= M_PI / 2.0)
            break;
        ++NP_on_hemisphere;
    }

    return NP_on_hemisphere;
}

static bool
SaveToFile(const char *filename, bool hemisphere)
{
    Tvector3 p;
    int ith_point;
    int last = (hemisphere ? NP_on_hemisphere : Number_of_points);

    bool ret = false;
    if (FILE *fp = fopen(filename, "w"))
    {
        for (ith_point = 1; ith_point <= last; ++ith_point)
        {
            p = Spherical2Cartesian(GSS[ith_point]);
            fprintf(fp, "%g\t%g\t%g\n", p.x, p.y, -p.z);
        }
        fclose(fp);
        ret = true;
    }
    else
    {
        fprintf(stderr, "SphereDist: Cannot open file '%s'.\n", filename);
    }

    return ret;
}

bool SphereDist(std::vector<Tvector3>& vertexes, int N, bool hemisphere, bool relocate)
{
    if (!GenerateGSS(N))
        return false;

    if (relocate)
        RelocatePoints();

    GetNP_on_hemisphere();

    Tvector3 p;
    int ith_point;
    int last = (hemisphere ? NP_on_hemisphere : Number_of_points);
    for (ith_point = 1; ith_point <= last; ++ith_point)
    {
        p = Spherical2Cartesian(GSS[ith_point]);
        vertexes.push_back(p);
    }

    return true;
}

bool SphereDist(const char *filename, int N, bool hemisphere, bool relocate)
{
    if (!GenerateGSS(N))
        return false;

    if (relocate)
        RelocatePoints();

    GetNP_on_hemisphere();

    return SaveToFile(filename, hemisphere);
}

File: SphereDist.js
// in radians
function TSpherical_Coordinates(col = 0.0, lon = 0.0) {
    this.col = col; // colatitude
    this.lon = lon; // longitude
}

function Tvector2(x = 0.0, y = 0.0) {
    this.x = x;
    this.y = y;
}

function Tvector3(x = 0.0, y = 0.0, z = 0.0) {
    this.x = x;
    this.y = y;
    this.z = z;
    this.add = function(another) {
        this.x += another.x;
        this.y += another.y;
        this.z += another.z;
    }
}

function sqr(x) {
    return x * x;
}

function Spherical2Cartesian(p) {
    return new Tvector3(
        Math.sin(p.col) * Math.cos(p.lon),
        Math.sin(p.col) * Math.sin(p.lon),
        Math.cos(p.col));
}

function Cartesian2Spherical(p) {
    return new TSpherical_Coordinates(
        Math.acos(p.z / Math.sqrt(sqr(p.x) + sqr(p.y) + sqr(p.z))),
        Math.atan2(p.y, p.x));
}

function SphereDist(N = 100, hemisphere = false, relocating = true) {
    this.version = "SphereDist version 0.8.3 by katahiromz";
    this.N = N;
    this.NP_on_hemisphere = N / 2;
    this.hemisphere = hemisphere;
    this.relocating = relocating;
    this.GSS = [];

    this.GetNP_on_hemisphere = function() {
        var NP_on_hemisphere = 0;
        for (var ith_point = 1; ith_point <= this.N; ++ith_point) {
            if (this.GSS[ith_point].col <= Math.PI / 2.0)
                break;
            ++NP_on_hemisphere;
        }
        this.NP_on_hemisphere = NP_on_hemisphere;
        return NP_on_hemisphere;
    }

    this.relocate = function() {
        const five_points = [0xFFFF, 2, 3, 5, 6, 7];

        // relocate 1st point
        var mean = new Tvector3();
        for (var i = 1; i <= 5; ++i) {
            var m = five_points[i];
            mean.add(Spherical2Cartesian(this.GSS[m]));
        }
        this.GSS[1] = Cartesian2Spherical(mean);

        // relocate Nth point
        mean = new Tvector3();
        for (var i = 1; i <= 5; ++i) {
            var m = five_points[i];
            mean.add(Spherical2Cartesian(this.GSS[this.N - m + 1]));
        }
        this.GSS[this.N] = Cartesian2Spherical(mean);
    }

    this.generate = function() {
        if (this.N < 6) {
            throw "N must be >= 6";
        }

        this.GSS = [];
        for (var i = 0; i < this.N + 1; ++i) {
            this.GSS.push(new TSpherical_Coordinates());
        }

        this.GSS[1].lon = 0.0;
        this.GSS[1].col = Math.PI;

        for (var k = 2; k <= this.N - 1; ++k) {
            var hk = -1.0 + 2.0 * (k - 1) / (this.N - 1);
            this.GSS[k].col = Math.acos(hk);
            this.GSS[k].lon = 
                this.GSS[k - 1].lon +
                    3.6 / Math.sqrt(this.N) / Math.sqrt(1.0 - hk * hk);
        }

        this.GSS[this.N].lon = 0.0;
        this.GSS[this.N].col = 0.0;

        if (this.hemisphere) {
            this.GetNP_on_hemisphere();
        }

        if (this.relocating) {
            this.relocate();
        }
    }

    this.rows = function() {
        var vertexes = [];
        const last = this.hemisphere ? this.NP_on_hemisphere : this.N;
        for (var ith_point = 1; ith_point <= last; ++ith_point) {
            var p = Spherical2Cartesian(this.GSS[ith_point]);
            vertexes.push([p.x, p.y, -p.z]);
        }
        return vertexes;
    }

    this.get_text = function() {
        var vertexes = this.rows();
        var text = "";
        for (var i = 0; i < vertexes.length; ++i) {
            text += vertexes[i][0];
            text += ", ";
            text += vertexes[i][1];
            text += ", ";
            text += vertexes[i][2];
            text += "\n";
        }
        return text;
    }

    this.generate();
}
